### PR TITLE
Bump Selenium Grid Version

### PR DESCRIPTION
**Pull request by ardelato**
## Description

This change upgrades the Selenium Grid from approximately version 3.14 to around 4.16.

To implement this upgrade, a TOML configuration file is required. This file is essential for specifying the browsers available on the grid and their corresponding drivers. This setup us to ensure that when there's a request for a new session, the `Distributor` component can accurately match the request with browsers that match the "`browserName`" and "`browserVersion`". The upgrade modified the way Selenium pairs browsers, moving away from matching solely based on "`browserName`". Now, the pairing process also considers "`browserVersion`" and "`platformName`".

Reference: https://ifixit.slack.com/archives/C01DBJK0H7U/p1705535705285359

### CR

There will be a follow-up in the `iFixit/ifixit` repo to drop the requirement of a browser version when sending in a session request. Our PHP tests do not cover different browser versions, so this request option is no longer needed. 

### QA

- [ ] Confirm you can run tests with the new local Selenium Grid.

1. Follow the instructions in https://overlord.ifixit.com/Guide/How+to+setup+Selenium+locally/786
2. Check out this branch on Cominor - https://github.com/iFixit/ifixit/pull/51703
3. Run a test on Cominor
	- Example: `platformName=MAC test-suite -sm GuideSmokeTest`

Closes: https://github.com/iFixit/ifixit/issues/51629

### CODE REVIEW

File: selenium-config.template.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[distributor]
+reject-unsupported-caps = true # Reject a request immediately if the Grid does not support the requested capability.
+
+[node]
+detect-drivers = false # Ignore local drivers and use the ones specified in the driver configuration section.
+
+[[node.driver-configuration]]
+display-name = "chrome"
+max-sessions = 2
```
Comment (mlahargou) [inline selenium-config.template.toml:9]: Let's increase this. I sometimes use many browsers to try to deflake tests. It's just the max, so maybe 20?

File: selenium-config.template.toml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+webdriver-path = "./chromedriver"
+stereotype = '{"browserName": "chrome", "browserVersion": "v120.0"}' # Browser version is not used by the Grid, but it is used by the tests to determine which browser to use.
```
Comment (mlahargou) [inline selenium-config.template.toml:11]: Hmm, if the user's local chrome version is different than this, will things break?

**Review by mlahargou — CHANGES_REQUESTED** [review]
dev_block 📱

File: selenium-config.template.toml (hunk 2 above)
Comment (mlahargou) [reply selenium-config.template.toml:11]: Oh, I think this should be
```suggestion
stereotype = '{"browserName": "chrome", "browserVersion": "v$full_chrome_version"}' # Browser version is not used by the Grid, but it is used by the tests to determine which browser to use.
```

or something like that

Comment (ardelato) [reply selenium-config.template.toml]: Addressed in [bc4c42f](https://github.com/iFixit/selenium_setup/pull/17/commits/bc4c42f5a25ad7b86315f20b71ff634a8fd19896)

File: selenium-config.template.toml (hunk 1 above)
Comment (ardelato) [reply selenium-config.template.toml]: Address in [`4fbe4a6` (#17)](https://github.com/iFixit/selenium_setup/pull/17/commits/4fbe4a645bf34329a3c99f730202cd9ed8a426e4)

**Comment by ardelato** [comment]
un_dev_block ✍️

**Review by mlahargou — APPROVED** [review]
CR 📱 Looks good to me. QA will be the real test.

**Comment by deltuh-vee** [comment]
dev_block ✊ 
I'm getting this error when running Selenium tests.
image

**Comment by mlahargou** [comment]
Try adding `version=120.0` (or whatever chrome version you run locally) to the `test-suite` command.

**Comment by ardelato** [comment]
Paired up with @deltuh-vee and there seems to be something else going on. That said, by dropping the browser version parameter from the request and in the Selenium Grid stereotype the tests were able to run.

However, since that solution required a bit of modification in both the `iFixit/ifixit` repo and this pull, I am going to open the follow-up now and have these two pulls be a pair.

**Comment by ardelato** [comment]
un_dev_block ⚡ I opened the corresponding pull in `iFixit/ifixit` which should help to avoid the whole `browserVersion` configuration step.

**Review by mlahargou — APPROVED** [review]
CR 📱 Cool!

**Comment by kthaler** [comment]
dev_block 🦖  - I’m having trouble successfully running the tests. I’ve followed the steps, and can successfully launch selenium:

```
> selenium_setup % ./selenium                                      
> 
> Forwarding port: remote:4428 to localhost:4444
> 
> Launching Selenium...
> 14:30:25.998 INFO [GridLauncherV3.parse] - Selenium server version: 3.141.59, revision: e82be7d358
> 14:30:26.087 INFO [GridLauncherV3.lambda$buildLaunchers$3] - Launching a standalone Selenium Server on port 4444
> 2024-01-22 14:30:26.148:INFO::main: Logging initialized @386ms to org.seleniumhq.jetty9.util.log.StdErrLog
> 14:30:26.437 INFO [WebDriverServlet.<init>] - Initialising WebDriverServlet
> 14:30:26.567 INFO [SeleniumServer.boot] - Selenium Server is up and running on port 4444

```

And then I run `echo 4428 > ~/.selenium_port`  in cominor on [this branch](https://github.com/iFixit/ifixit/pull/51703). But when I run the test-suite, I get a bunch of fails displaying: `Failed to connect to 127.0.0.1 port 4428 after 0 ms: Couldn't connect to server`

<details>
<summary>Here is the full error</summary>
<br>
WDGuideSmokeTest error: 

        PHPUnit 10.5.2 by Sebastian Bergmann and contributors.

        Runtime:       PHP 8.2.8
        Configuration: /mnt/ebs/home/kthaler/Code/Tests/phpunit.xml

        EEEEE                                                               5 / 5 (100%)

        Time: 00:00.499, Memory: 20.00 MB

        There were 5 errors:

        1) WDGuideSmokeTest::testCreate
        Facebook\WebDriver\Exception\Internal\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"capabilities":{"firstMatch":[{"browserName":"chrome","platformName":"mac","acceptInsecureCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}]},"desiredCapabilities":{"browserName":"chrome","platform":"MAC","acceptSslCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}}

        Recv failure: Connection reset by peer

        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php:20
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php:328
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:129
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriver.php:41
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:7
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:34
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UITestCase.php:10

        2) WDGuideSmokeTest::testGuideCompletion
        Facebook\WebDriver\Exception\Internal\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"capabilities":{"firstMatch":[{"browserName":"chrome","platformName":"mac","acceptInsecureCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}]},"desiredCapabilities":{"browserName":"chrome","platform":"MAC","acceptSslCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}}

        Failed to connect to 127.0.0.1 port 4428 after 0 ms: Couldn't connect to server

        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php:20
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php:328
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:129
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriver.php:41
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:7
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:34
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UITestCase.php:10

        3) WDGuideSmokeTest::testGuideEdits
        Facebook\WebDriver\Exception\Internal\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"capabilities":{"firstMatch":[{"browserName":"chrome","platformName":"mac","acceptInsecureCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}]},"desiredCapabilities":{"browserName":"chrome","platform":"MAC","acceptSslCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}}

        Failed to connect to 127.0.0.1 port 4428 after 0 ms: Couldn't connect to server

        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php:20
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php:328
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:129
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriver.php:41
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:7
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:34
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UITestCase.php:10

        4) WDGuideSmokeTest::testCreatePlaintext
        Facebook\WebDriver\Exception\Internal\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"capabilities":{"firstMatch":[{"browserName":"chrome","platformName":"mac","acceptInsecureCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}]},"desiredCapabilities":{"browserName":"chrome","platform":"MAC","acceptSslCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}}

        Failed to connect to 127.0.0.1 port 4428 after 0 ms: Couldn't connect to server

        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php:20
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php:328
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:129
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriver.php:41
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:7
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:34
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UITestCase.php:10

        5) WDGuideSmokeTest::testEditPlaintext
        Facebook\WebDriver\Exception\Internal\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"capabilities":{"firstMatch":[{"browserName":"chrome","platformName":"mac","acceptInsecureCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}]},"desiredCapabilities":{"browserName":"chrome","platform":"MAC","acceptSslCerts":true,"goog:chromeOptions":{"args":["--disk-cache-dir=/tmp/","--hide-scrollbars"]}}}

        Failed to connect to 127.0.0.1 port 4428 after 0 ms: Couldn't connect to server

        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php:20
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php:328
        /mnt/ebs/home/kthaler/Code/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:129
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriver.php:41
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:7
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UIDriverFactory.php:34
        /mnt/ebs/home/kthaler/Code/Tests/Objects/UITestCase.php:10

        ERRORS!
        Tests: 5, Assertions: 0, Errors: 5.

        Command exit status was 2 and killing signal was null

-------------------------
Tests that failed 1 times
-------------------------
(00:00.60) - WDGuideSmokeTest

Rerun the tests that failed 1 times with this command:
host=amazon test-suite -s WDGuideSmokeTest
</details>

**Comment by mlahargou** [comment]
> I’m having trouble successfully running tests.

Looks like you are running the wrong selenium server version:
`Selenium server version: 3.141.59`

**Comment by kthaler** [comment]
> Are you sure you have this branch checked out?

I did not, but after redoing the steps with this branch, I'm getting the same result. Here's the logs when I'm launching selenium on this branch:

```
> Launching Selenium...
> 16:41:52.009 INFO [LoggingOptions.configureLogEncoding] - Using the system default encoding
> 16:41:52.032 INFO [OpenTelemetryTracer.createTracer] - Using OpenTelemetry for tracing
> 16:41:53.190 INFO [NodeOptions.getSessionFactories] - Detected 10 available processors
> 16:41:53.227 INFO [NodeOptions.report] - Adding chrome for {"browserName": "chrome","platformName": "mac"} 10 times
> 16:41:53.276 INFO [Node.<init>] - Binding additional locator mechanisms: relative
> 16:41:53.301 INFO [GridModel.setAvailability] - Switching Node 5f1d5ade-1018-4074-9ccd-f1ccf4fdfa75 (uri: http://192.168.1.11:4444) from DOWN to UP
> 16:41:53.302 INFO [LocalDistributor.add] - Added node 5f1d5ade-1018-4074-9ccd-f1ccf4fdfa75 at http://192.168.1.11:4444. Health check every 120s
> 16:41:53.416 INFO [Standalone.execute] - Started Selenium Standalone 4.16.1 (revision 9b4c83354e): http://192.168.1.11:4444

```

**Comment by ardelato** [comment]
Paired up with Curtis again, there seems to be other factors which could leave to inconsistent behavior such as if there is a `~/.cache/selenium` folder and possible the Java version that is currently installed.

@mlahargou what are your thoughts on punting the update to the local Selenium version for now and work on maybe getting a containerized version of local selenium?

**Comment by mlahargou** [comment]
QA 📱 

Same as: https://github.com/iFixit/ifixit/pull/51703#issuecomment-1906643810

**Pull request merged**